File: src/main/java/kz/ncanode/configuration/JacksonConfiguration.java
package kz.ncanode.configuration;

import org.springframework.context.annotation.Configuration;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

import java.util.List;

/**
 * Регистрирует Jackson 2 как первый JSON HttpMessageConverter.
 *
 * Spring Boot 4 авто-конфигурит Jackson 3 (tools.jackson.core) как default
 * ObjectMapper. Существующие Java DTO (CmsVerifyRequest и др.) используют
 * Lombok @Jacksonized, который генерирует Jackson 2 аннотации
 * (com.fasterxml.jackson.databind.annotation.JsonDeserialize). Jackson 3
 * эти аннотации не распознаёт — получается 500 "Type definition error".
 *
 * Spring Framework 7 сохранил MappingJackson2HttpMessageConverter для
 * обратной совместимости — регистрируем его на индекс 0, чтобы он обрабатывал
 * application/json первее Jackson 3'шного.
 *
 * Уйдёт когда все DTO будут портированы в Kotlin data classes (Jackson 3
 * понимает их через kotlin-module нативно).
 */
@Configuration
public class JacksonConfiguration implements WebMvcConfigurer {

    @Override
    public void configureMessageConverters(List<HttpMessageConverter<?>> converters) {
        converters.add(0, new MappingJackson2HttpMessageConverter());
    }
}

File: src/main/kotlin/kz/ncanode/configuration/JacksonConfiguration.kt
package kz.ncanode.configuration

import org.springframework.context.annotation.Configuration
import org.springframework.http.MediaType
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer

/**
 * Регистрирует Jackson 2 как первый JSON HttpMessageConverter — но
 * ограниченный нашими DTO-классами (`kz.ncanode.dto.*`).
 *
 * Spring Boot 4 авто-конфигурит Jackson 3 (tools.jackson.core) как default
 * ObjectMapper. Существующие Java DTO используют Lombok `@Jacksonized`,
 * который генерирует Jackson 2 аннотации (com.fasterxml.jackson.*).
 * Jackson 3 эти аннотации не распознаёт → 500 "Type definition error".
 *
 * Поэтому Jackson 2 обрабатывает только наши DTO; всё остальное
 * (byte[] от springdoc, String, статика) идёт через default-конвертеры
 * Spring Boot. Без `isOurDto` фильтра Jackson 2 перехватывал бы byte[]
 * от `/v3/api-docs` и кодировал его в base64.
 *
 * Уйдёт целиком когда все DTO будут портированы в Kotlin data classes
 * (Jackson 3 поймёт их нативно через jackson-module-kotlin).
 */
@Configuration
@Suppress("removal", "DEPRECATION")
class JacksonConfiguration : WebMvcConfigurer {

    override fun extendMessageConverters(converters: MutableList<HttpMessageConverter<*>>) {
        converters.add(0, NcanodeJackson2Converter())
    }

    private class NcanodeJackson2Converter : MappingJackson2HttpMessageConverter() {
        private fun isOurDto(clazz: Class<*>): Boolean =
            clazz.name.startsWith("kz.ncanode.dto.")

        override fun canRead(clazz: Class<*>, mediaType: MediaType?): Boolean =
            isOurDto(clazz) && super.canRead(clazz, mediaType)

        override fun canWrite(clazz: Class<*>, mediaType: MediaType?): Boolean =
            isOurDto(clazz) && super.canWrite(clazz, mediaType)
    }
}
